samples/msal-node-samples/nextjs-rsc-ts/src/utils/AuthProvider.ts

```diff
@@ -0,0 +1,174 @@
+  Configuration,
+  CryptoProvider,
+  DistributedCachePlugin,
+  ICacheClient,
+  IPartitionManager,
+} from "@azure/msal-node";
+import { cache } from "react";
+
+export type PartitionManagerFactory = () => Promise<IPartitionManager>;
+
+type AuthCodeRequestState = {
+  returnTo: string;
+  request: AuthorizationCodeRequest;
+};
+
+/**
+ * Light wrapper for msal-node's ConfidentialClientApplication.
+ */
+export class AuthProvider {
+  configuration: Configuration;
+  cacheClient: ICacheClient;
+  partitionManagerFactory: PartitionManagerFactory;
+  cryptoProvider: CryptoProvider;
+  redirectUri: string;
+
+  /**
+   * Initialize the AuthProvider.
+   * @param configuration The msal configuration object
+   * @param redirectUri Uri that authentication requests will redirect back to
+   * @param cacheClient The cache client used to store the token cache
+   * @param partitionManagerFactory Factory that returns a PartitionManager for the current executing context
+   */
+  constructor(
+    configuration: Configuration,
+    redirectUri: string,
+    cacheClient: ICacheClient,
+    partitionManagerFactory: PartitionManagerFactory
+  ) {
+    this.configuration = configuration;
+    this.redirectUri = redirectUri;
+    this.cacheClient = cacheClient;
+    this.partitionManagerFactory = cache(partitionManagerFactory);
+
+    this.cryptoProvider = new CryptoProvider();
+  }
+
+  /**
+   * Get url for an auth code request
+   * @param request Authorization request to initialize the flow
+   * @param returnTo Where the user should be redirected to after a successful flow
+   * @returns The url to redirect the client to
+   */
+  async getAuthCodeUrl(
+    request: Omit<AuthorizationUrlRequest, "redirectUri">,
+    returnTo: string
+  ) {
+    const instance = await this.getInstance();
+
+    const requestWithRedirectUri = {
+      ...request,
+      redirectUri: this.redirectUri,
+    };
+
+    const state = this.cryptoProvider.base64Encode(
+      JSON.stringify({
+        request: requestWithRedirectUri,
+        returnTo,
+      })
+    );
+
+    return await instance.getAuthCodeUrl({
+      ...requestWithRedirectUri,
+      state: state,
+    });
+  }
+
+  /**
+   * Handles token acquisition based on the url that the user was sent to from Azure.
+   * @param url The return url from Azure
+   * @returns An object containing the logged in account, and where the user should be redirected to.
+   */
+  async handleAuthCodeCallback(url: URL) {
+    const instance = await this.getInstance();
+
+    const stateString = url.searchParams.get("state");
+
+    if (!stateString) {
+      throw new Error("No state found.");
+    }
+
+    const state: AuthCodeRequestState = JSON.parse(
+      this.cryptoProvider.base64Decode(stateString)
+    );
+
+    const code = url.searchParams.get("code");
+
+    if (!code) {
+      throw new Error("No code found.");
+    }
+
+    const authResult = await instance.acquireTokenByCode({
+      ...state.request,
+      code,
```

Comment:
you can pass the entire query response into a JS object, so that acquireTokenByCode also validates the returned parameters ([see](https://azuread.github.io/microsoft-authentication-library-for-js/ref/classes/_azure_msal_node.confidentialclientapplication.html#acquiretokenbycode)). Something like:

```js
const urlParams = new URLSearchParams(url);
const params = Object.fromEntries(urlParams)

const authResult = await instance.acquireTokenByCode({
      ...state.request,
      params
});
```

Though I'd much setting the response mode to `form_post` in the authCodeUrl request, and then receiving the auth code response as a POST request to your server (recommended for confidential client apps in general).

Reply:
Switched to using `ResponseMode.FORM_POST`, and passing all form data to the second argument.